src/snowflake/snowpark/async_job.py

```diff
@@ -237,7 +239,19 @@ def is_done(self) -> bool:
     def cancel(self) -> None:
         """Cancels the query associated with this instance."""
         # stop and cancel current query id
-        self._cursor.execute(f"select SYSTEM$CANCEL_QUERY('{self.query_id}')")
+        if (
+            is_in_stored_procedure()
+            and self._session._conn._get_client_side_session_parameter(
+                "ENABLE_ASYNC_QUERY_IN_PYTHON_STORED_PROCS", False
+            )
+        ):
+            cancel_resp = self._session._conn._conn.cancel_query(self.query_id)
+            if "success" not in cancel_resp or not cancel_resp["success"]:
+                raise DatabaseError(
+                    f"Failed to cancel query. Returned response: {cancel_resp}"
+                )
+        else:
+            self._cursor.execute(f"select SYSTEM$CANCEL_QUERY('{self.query_id}')")
```

Comment:
I wonder if we could add a `cancel_query` API in python connector as well. This way we only need to call `cancel_query` from connector?

Reply:
I think we should. There seems to be no public API in the connector to cancel query.

Reply:
Agree we should. I created a separate jira (SNOW-941385) to add the API in public python connector.